### PR TITLE
dns: tie initial load of lookup table to readiness probe

### PR DESCRIPTION
**Pull request by ramaraochavali**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by howardjohn** [comment]
I think this may be the right thing to do, but one concern: a user sets up ServiceEntrywith address=1.2.3.4 for google.com. At startup, pod resolves google.com to 6.7.8.9 (some real google.com IP), which may be unexpected.

We have similar blocking by default behavior for non-DNS. For example, we do not setup a passthrough listener in bootstrap and allow requests through before envoy init, so I am not sure if we should do it for DNS?

**Comment by ramaraochavali** [comment]
I think we should let DNS forward to upstream if it does not have info for some reason (like NameTable may be initially populated but the new entry might not have been updated because Istiod -> Agent path is down or it takes time - in this case we actually any forward to upstream because host is not found ).  If people complain about it may be we can add a config to agent to control this behaviour. WDYT?

**Comment by howardjohn** [comment]
Why should DNS requests have this behavior when TCP requests do not?

> like NameTable may be initially populated but the new entry might not have been updated because Istiod -> Agent path is down or it takes time - in this case we actually any forward to upstream because host is not found

The initial XDS response will always be complete, if its not its a P0 bug. New SE may not apply immediately, but once it is applied all new pods need to get it in the first push

**Comment by ramaraochavali** [comment]
> Why should DNS requests have this behavior when TCP requests do not?

The reason I think it is needed is the we have enabled DNS_CAPTURE of K8s Jobs and the job makes a DNS query and because nametable is not loaded and DNS it self fails - which I think is unexpected. For TCP calls - it checks for proxy readiness before making calls. 
Either we allow DNS or make DNS also part of readiness check?

**Comment by howardjohn** [comment]
I am not sure I understand. DNS is only used for outbound calls. Proxy readiness only impacts inbound calls? Or does your job have some special logic to wait until envoy is ready? I suppose we could also get inbound calls that trigger outbound calls. I think adding DNS to readiness makes sense.

**Comment by ramaraochavali** [comment]
The job does a regular DNS query and it fails because we set the DNS_CAPTURE to true and our proxy returns failure.

Later it also uses envoy readiness before it uses mesh.

**Comment by howardjohn** [comment]
Ok, we should just add it to the readiness then? I don't think we can have any expectation of networking working before proxy is initialized

**Comment by ramaraochavali** [comment]
@howardjohn Added to readiness. PTAL

File: pilot/pkg/dns/dns.go [hunk 1]
```diff
@@ -153,6 +156,7 @@ func (h *LocalDNSServer) UpdateLookupTable(nt *nds.NameTable) {
 		lookupTable.buildDNSAnswers(altHosts, ipv4, ipv6, h.searchNamespaces)
 	}
 	h.lookupTable.Store(lookupTable)
+	h.initialized = true
```
Comment (howardjohn) [inline pilot/pkg/dns/dns.go:159]: Is this concurrency safe?

File: pilot/pkg/dns/dns.go [hunk 2]
```diff
@@ -215,16 +219,18 @@ func (h *LocalDNSServer) ServeDNS(proxy *dnsProxy, w dns.ResponseWriter, req *dn
...
 	}
+
 	// We did not find the host in our internal cache. Query upstream and return the response as is.
-	response = h.queryUpstream(proxy.upstreamClient, req, log)
-	// Compress the response - we don't know if the incoming response was compressed or not. If it was,
-	// but we don't compress on the outbound, we will run into issues. For example, if the compressed
-	// size is 450 bytes but uncompressed 1000 bytes now we are outside of the non-eDNS UDP size limits
-	response.Truncate(size(proxy.protocol, req))
 	log.Debugf("response for hostname %q (found=false): %v", hostname, response)
+	response = h.queryUpstream(proxy.upstreamClient, req, log)
```
Comment (howardjohn) [inline pilot/pkg/dns/dns.go:225]: why did we change the truncation?

File: pilot/pkg/dns/dns.go [hunk 3]
```diff
@@ -44,6 +44,8 @@ type LocalDNSServer struct {
 	// Optimizations to save space and time
 	proxyDomain      string
 	proxyDomainParts []string
+	// Stores if lookupTable is updated atleast once.
+	initialized bool
```
Comment (howardjohn) [inline pilot/pkg/dns/dns.go:48]: We could also check `lookupTable != nil` maybe?

File: pilot/pkg/dns/dns.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/dns/dns.go:225]: Let me check. I did not change intentionally

File: pilot/pkg/dns/dns.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/dns/dns.go:159]: UpdateLookUpTable is not concurrent right?

Comment (howardjohn) [reply pilot/pkg/dns/dns.go:159]: I was thinking it was but I suppose not. But UpdateLookUpTable may change the value while the probe checks it?

File: pilot/pkg/dns/dns.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/dns/dns.go]: We have to use h.lookupTable.Load() != nil - was trying to avoid Load call. But may be OK. Changed now. PTAL

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/dns/dns.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/dns/dns.go:222]: It is removed by mistake. Reverted it

**Pull request merged**